### ISSUE TITLE
SqlDataProvider with PostgreSQL DB

### ISSUE THREAD

**Issue by hromatka** [mvysny/vaadin-on-kotlin#19]
First, after 2 weeks of working with VoK, I am still fascinated by its elegancy. Very nice work! Converting my few small Vaadin projects to VoK, i ran into this problem:

Using SqlDataProvider to map the result of a Postgresql table valued function, I got a `Database error: Operation requires a scrollable ResultSet, but this ResultSet is FORWARD_ONLY.` I believe it is a known problem as I could see plenty of post suggesting to manually set `ResultSet.TYPE_SCROLL_INSENSITIVE`. I could track down the issue to:
```
override fun sizeInBackEnd(query: Query<T, Filter<T>?>?): Int
(...)
val counts: List<Int> = q.executeAndFetch({ rs: ResultSet -> if (rs.last()) rs.row else 0 })
```
Unfortunately I do not quiet understand how to set this up correctly, my (SQL guy) solution was to wrap the query into a `SELECT count(*) FROM ({$s}) AS Tab` in `computeSQL` and do the counting on the Server, but I realize that this could be DB dependent.

One footnote: the statement `... ORDER BY NULL{{Where}}` leads to an error on PostgreSQL, I used `... ORDER BY 1=1{{Where}}` which amazingly does work (you can not use ORDER BY 1 as this uses the first row for sorting).

Thanx again,

**Comment by mvysny** [comment]
> First, after 2 weeks of working with VoK, I am still fascinated by its elegancy. Very nice work!

Hey, thanks! I'm very happy that you like the simplicity :)

> Unfortunately I do not quiet understand how to set this up correctly, my (SQL guy) solution was to wrap the query into a SELECT count(*) FROM ({$s}) AS Tab in computeSQL and do the counting on the Server, but I realize that this could be DB dependent.

Actually I pretty much like the `SELECT count(*) FROM ({$s})` solution! Probably way more efficient than running the whole select and just scrolling to the bottom, to check the number of rows (the `rs.last()` part). I need to add the tests for all supported databases anyway (so far H2, MariaDB, MySQL and PostgreSQL), so I will optimize the `count` query as you suggested and let the tests tell how the databases support that :)

> One footnote: the statement ... ORDER BY NULL{{Where}} leads to an error on PostgreSQL, I used ... ORDER BY 1=1{{Where}} which amazingly does work (you can not use ORDER BY 1 as this uses the first row for sorting).

Thanks for letting me know, I will update the documentation for `SqlDataProvider` regarding this.

**Comment by mvysny** [comment]
With your permission I will use your trick - apparently it works with H2, MariaDB, MySQL and PostgreSQL so I'm happy :) I have added tests for all supported databases both for `SqlDataProvider` and `EntityDataProvider`.
Fixed in VoK 0.4.4
